### DISCUSSION TITLE
json-c (and possibly others) memory leaks?

### DISCUSSION THREAD

**Discussion by 50b5** [termux/termux-packages#11209] [Q&A]
I'm pretty sure this is an issue in Android's libc but I'm surprised at how little information (including from google's git site) I've found about it.

https://github.com/json-c/json-c/issues/668 and https://github.com/json-c/json-c/issues/705 explain the problem and I can confirm any time I use ``json_tokener_parse`` or the test program they wrote, it leaks memory.

I wasn't sure where to post this since I could just be missing something obvious but has anyone encountered this?

**Comment by 50b5** [comment]
So I did some digging.

```c++
locale_t newlocale(int category_mask, const char* locale_name, locale_t /*base*/) {
  // Are 'category_mask' and 'locale_name' valid?
  if ((category_mask & ~LC_ALL_MASK) != 0 || locale_name == nullptr) {
    errno = EINVAL;
    return nullptr;
  }

  if (!__is_supported_locale(locale_name)) {
    errno = ENOENT;
    return nullptr;
  }

  return __alloc_locale(__is_utf8_locale(locale_name) ? 4 : 1);
}
```

Taken from [bionic master](https://android.googlesource.com/platform/bionic/+/refs/heads/master/libc/bionic/locale.cpp) and turns out that it just ignores the base argument. So any program that tries to use newlocale as POSIX intended will leak memory. I'm bringing this up here since I don't know if json-c is the only package under termux that uses newlocale this way. Any thoughts?

Sorry for the bump btw. Just trying to figure out what to do about this.

**Reply by licy183** [reply]
Disable `locale` in `json-c` in #11220.

> So any program that tries to use newlocale as POSIX intended will leak memory. I'm bringing this up here since I don't know if json-c is the only package under termux that uses newlocale this way. Any thoughts?

Yeah, I think it is difficult and almost impossible for us to find out all the problematic packages. Unless user reports these questions I think we have no way to know...

**Reply by 50b5** [reply]
Thanks! Wow. Would it be possible to get bionic devs to change that? I don't see a reason why that argument is just ignored but I suppose if they built around it that way it could be a pretty major change they might not accept. I guess that doesn't even include how long it'd take to roll out...

Well I'm new to this side of GitHub but I can look through some more packages and try to help out. Thanks again!

**Reply by licy183** [reply]
> Thanks! Wow. Would it be possible to get bionic devs to change that? I don't see a reason why that argument is just ignored but I suppose if they built around it that way it could be a pretty major change they might not accept. I guess that doesn't even include how long it'd take to roll out...

Well, I think this problem is not really related to `bionic libc`. As stated in the Linux manual page of [duplocale](https://man7.org/linux/man-pages/man3/duplocale.3.html) and [newlocale](https://man7.org/linux/man-pages/man3/newlocale.3.html), `Each locale object created by duplocale()/newlocale() should be deallocated
 using freelocale()`. So I think it is an upstream issue of `json-c`. 

> Well I'm new to this side of GitHub but I can look through some more packages and try to help out. Thanks again!

Welcome to submit Pull Request or Issues if you find other packages in Termux using `duplocale` without `freelocale`!
